src/libraries/Microsoft.PowerFx.Core/Localization/StringResources.cs

```diff
@@ -25,72 +24,193 @@ internal static class StringResources
+                {
+                    str = StringResources.ExternalStringResources.TryGet(resourceKey, out string resourceValue, locale) ? resourceValue : null;
+
+                    if (!string.IsNullOrEmpty(str))
+                    {
+                        return str;
+                    }
+                }
+
+                ThrowInternal(resourceKey);
+            }
+
+            return Get(resourceKey, locale, true);
+        }
+
+        internal string Get(string resourceKey, string locale = null, bool throwIfNotFound = true)
+        {
+            Contracts.CheckValue(resourceKey, nameof(resourceKey));
+            Contracts.CheckValueOrNull(locale, nameof(locale));
+
+            if (!((IExternalStringResources)this).TryGet(resourceKey, out var resourceValue, locale))
             {
                 // Prior to ErrorResources, error messages were fetched like other string resources.
                 // The resource associated with the key corresponds to the ShortMessage of the new
                 // ErrorResource objects. For backwards compatibility with tests/telemetry that fetched
                 // the error message manually (as opposed to going through the DocError class), we check
                 // if there is an error resource associated with this key if we did not find it normally.
-                if (TryGetErrorResource(new ErrorResourceKey(resourceKey), out var potentialErrorResource, locale))
+                if (((IExternalStringResources)this).TryGetErrorResource(new ErrorResourceKey(resourceKey, this), out var potentialErrorResource, locale))
                 {
                     return potentialErrorResource.GetSingleValue(ErrorResource.ShortMessageTag);
                 }
 
-                Debug.WriteLine(string.Format(CultureInfo.InvariantCulture, "ERROR resource string {0} not found", resourceKey));
-                if (ShouldThrowIfMissing)
+                if (throwIfNotFound)
                 {
-                    throw new System.IO.FileNotFoundException(resourceKey);
+                    ThrowInternal(resourceKey);
                 }
+
+                return null;
             }
 
             return resourceValue;
-        }        
+        }
 
-        public static bool TryGet(string resourceKey, out string resourceValue, string locale = null)
+        internal static void ThrowInternal(string resourceKey)
         {
-            Contracts.CheckValue(resourceKey, nameof(resourceKey));
-            Contracts.CheckValueOrNull(locale, nameof(locale));
-
-            resourceValue = _resourceManager.GetLocaleResource(resourceKey, locale);
-
-            return resourceValue != null ? true : (ExternalStringResources?.TryGet(resourceKey, out resourceValue, locale) ?? false);
+            Debug.WriteLine(string.Format(CultureInfo.InvariantCulture, "ERROR resource string {0} not found", resourceKey));
+            throw new System.IO.FileNotFoundException(resourceKey);
```

Comment:
>FileNotFoundException [](http://example.com/codeflow?start=32&length=21)

It's not really a file not found exception; it could be that the resource file is there, but the resource is not. KeyNotFoundException?

Reply:
I prefer keeping this existing logic